### PR TITLE
fix: VMware Workstation installed to non-default path

### PR DESCRIPTION
**Pull request by sparshev**
Non-default directory (like `/usr/local`) will not work for packer properly due to the absolute paths are used in version detection. This change should fix this issue and allow the packer users to use any path for distributive they want.

As a workaround for current packer 1.8.1 - create this script to fix an issue until the patch will be applied and new packer version issued:
* `/usr/lib/vmware/bin/vmware-vmx` (executable):
   ```sh
   #!/bin/sh
   vmware -v 1>&2
   ```

### CODE REVIEW

File: builder/vmware/common/driver_workstation_unix.go
Comment (lbajolet-hashicorp) [inline builder/vmware/common/driver_workstation_unix.go:120]: Could you add a small comment to state that `appPath` is expected to point to the `vmware` tool installed on the machine?
That'd make it easier to understand the process of inferring the lib path.

The implementation looks OK to me otherwise

**Review by lbajolet-hashicorp — COMMENTED** [review]
The code looks OK to me, I'd wait for @nywilken's review if any, as he'll be more knowledgeable than me on this code to see if there's an implication to changing the way we get the paths to the different tools.

Thanks for the contribution!

File: builder/vmware/common/driver_workstation_unix.go
Comment (sparshev) [reply builder/vmware/common/driver_workstation_unix.go:130]: I think we're quite safe here, vmware executable is required by the [workstation driver](https://github.com/hashicorp/packer-plugin-vmware/blob/d4fc272/builder/vmware/common/driver_workstation9.go#L143) and if not found than it will not work at all. The `vmware` gives a bit different version than `vmware-vmx`, but at least the same:
* `vmware-vmx -v`
   ```
   $ /usr/local/lib/vmware/bin/vmware-vmx -v
   
   VMware Workstation Information:
   VMware Workstation 16.2.3 build-19376536 Release
   
   ```
* `vmware -v`:
   ```
   $ vmware -v
   VMware Workstation 16.2.3 build-19376536
   ```

Not sure how about the previous versions though... Right now the original (1.8.1) logic works for me with a workaround I described above.

File: builder/vmware/common/driver_workstation_unix.go
Comment (sparshev) [reply builder/vmware/common/driver_workstation_unix.go:122]: Ok, added here: f108aed

File: builder/vmware/common/driver_workstation_unix.go
Comment (sparshev) [reply builder/vmware/common/driver_workstation_unix.go:132]: Actually for additional safety I can leave the old logic as plan B in case the execution of appPath or matching of version will not work - so everyone will be happy, wdyt?

File: builder/vmware/common/driver_workstation_unix.go
Comment (lbajolet-hashicorp) [reply builder/vmware/common/driver_workstation_unix.go:122]: Perfect, thanks

File: builder/vmware/common/driver_workstation_unix.go
Comment (lbajolet-hashicorp) [reply builder/vmware/common/driver_workstation_unix.go:132]: With the output you provide and the version logic (it only takes the major version), I'd think we're safe with the implementation you propose.
No need for a plan B I'd think, I was probably just a bit overcautious.

**Review by lbajolet-hashicorp — APPROVED** [review]

**Review by nywilken — APPROVED** [review]
Change wise this looks good to me as well. I see the fusion driver has the ability to set a custom path, which is not true for the workstation driver. Maybe something to consider in the future. 

@sparshev @lbajolet-hashicorp when running `make testacc` do all test pass for you? 

I'm in agreement with you @lbajolet-hashicorp, if the the acceptance test fail we should take a look at the logs to make sure there is nothing being missed. But I don't suspect there will be any issues looking at the changes. It is more of a sanity check. Approving in advance.

**Review by nywilken — CHANGES_REQUESTED** [review]
👋 I spent a little time looking at the code to get a sense of the AppPath change. It looks like AppPath for the driver represents the path to the vmware executable, on my Linux distro with a stock install of VMware Workstation being `/usr/bin/vmware`. 

Given that this is the path to the executable this call 
```libPath := filepath.Dir(filepath.Dir(appPath))``` fails because the required tools are not under `/usr/bin/vmware/isoimages/*.iso` but under `/usr/lib/vmware/...`. See the error below when building with these changes:

I think the path forward here would be to revisit how AppPath gets set for the workstation drivers and making it customizable for both the executable and vmware-vmx path. 


Failure with latest changes. 
```
~> cd ~/packer-plugin-vmware-main/example/
~>  packer build .
vmware-iso.example: output will be in this color.

Build 'vmware-iso.example' errored after 1 second 572 milliseconds: Couldn't find VMware tools for 'linux'! VMware often downloads these
tools on-demand. However, to do this, you need to create a fake VM
of the proper type then click the 'install tools' option in the
VMware GUI.

==> Wait completed after 1 second 572 milliseconds

==> Some builds didn't complete successfully and had errors:
--> vmware-iso.example: Couldn't find VMware tools for 'linux'! VMware often downloads these
tools on-demand. However, to do this, you need to create a fake VM
of the proper type then click the 'install tools' option in the
VMware GUI.

==> Builds finished but no artifacts were created.
```

Reverting the change results in a build with no error (minus ssh failure)
```
~>  packer build .
vmware-iso.example: output will be in this color.

==> vmware-iso.example: Retrieving ISO
==> vmware-iso.example: Trying http://old-releases.ubuntu.com/releases/14.04.1/ubuntu-14.04.1-server-amd64.iso
==> vmware-iso.example: Trying http://old-releases.ubuntu.com/releases/14.04.1/ubuntu-14.04.1-server-amd64.iso?checksum=sha256%3A946a6077af6f5f95a51f82fdc44051c7aa19f9cfc5f737954845a6050543d7c2
==> vmware-iso.example: http://old-releases.ubuntu.com/releases/14.04.1/ubuntu-14.04.1-server-amd64.iso?checksum=sha256%3A946a6077af6f5f95a51f82fdc44051c7aa19f9cfc5f737954845a6050543d7c2 => /home/wilken/.cache/packer/0bc58f084cbec97ab962909e9e03eb3572b77174.iso
==> vmware-iso.example: Configuring output and export directories...
==> vmware-iso.example: Creating required virtual machine disks
==> vmware-iso.example: Building and writing VMX file
==> vmware-iso.example: Starting HTTP server on port 8658
==> vmware-iso.example: Starting virtual machine...
    vmware-iso.example: The VM will be run headless, without a GUI. If you want to
    vmware-iso.example: view the screen of the VM, connect via VNC with the password "pJ33Xs0o" to
    vmware-iso.example: vnc://127.0.0.1:5999
==> vmware-iso.example: Connecting to VNC...
```

**Comment by sparshev** [comment]
@nywilken you right! It's the thing I forget to test - I even named it libPath, but forget to join with lib directory. Thank you for finding that, I will execute the testacc today.

File: builder/vmware/common/driver_workstation_unix.go
Comment (sparshev) [reply builder/vmware/common/driver_workstation_unix.go:133]: Yep, correct - I marked that in workaround I provided

**Comment by sparshev** [comment]
Ok, finally ran the tests, seems they failed just in this section:
```
=== RUN   TestAccStepCreateVmx_SerialFile
    pluginacc.go:143: Error running plugin acceptance tests: Bad exit code. Logfile: packer_log_vmware-iso_create_vmx_serial_file.txt
        Logs can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/packer_log_vmware-iso_create_vmx_serial_file.txt
        and the acceptance test template can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/vmware-iso_create_vmx_serial_file.yaml
--- FAIL: TestAccStepCreateVmx_SerialFile (95.30s)
=== RUN   TestStepCreateVmx_SerialPort
    pluginacc.go:143: Error running plugin acceptance tests: Bad exit code. Logfile: packer_log_vmware-iso_create_vmx_serial_port.txt
        Logs can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/packer_log_vmware-iso_create_vmx_serial_port.txt
        and the acceptance test template can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/vmware-iso_create_vmx_serial_port.yaml
--- FAIL: TestStepCreateVmx_SerialPort (92.79s)
=== RUN   TestStepCreateVmx_ParallelPort
    pluginacc.go:143: Error running plugin acceptance tests: Bad exit code. Logfile: packer_log_vmware-iso_create_vmx_parallel_port.txt
        Logs can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/packer_log_vmware-iso_create_vmx_parallel_port.txt
        and the acceptance test template can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/vmware-iso_create_vmx_parallel_port.yaml
--- FAIL: TestStepCreateVmx_ParallelPort (92.84s)
=== RUN   TestStepCreateVmx_Usb
    pluginacc.go:143: Error running plugin acceptance tests: Bad exit code. Logfile: packer_log_vmware-iso_create_vmx_usb.txt
        Logs can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/packer_log_vmware-iso_create_vmx_usb.txt
        and the acceptance test template can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/vmware-iso_create_vmx_usb.yaml
--- FAIL: TestStepCreateVmx_Usb (92.88s)
=== RUN   TestStepCreateVmx_Sound
    pluginacc.go:143: Error running plugin acceptance tests: Bad exit code. Logfile: packer_log_vmware-iso_create_vmx_sound.txt
        Logs can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/packer_log_vmware-iso_create_vmx_sound.txt
        and the acceptance test template can be found at /home/user/git/packer-plugin-vmware/builder/vmware/iso/vmware-iso_create_vmx_sound.yaml
--- FAIL: TestStepCreateVmx_Sound (92.77s)
FAIL
FAIL    github.com/hashicorp/packer-plugin-vmware/builder/vmware/iso    840.490s
```

**Comment by nywilken** [comment]
Hi @sparshev apologies for the delayed response here. In talking about the change internally we ultimately landed on the idea that there should be a configuration argument for changing the installation path as opposed to trying to derive it within Packer. The Fusion driver has a similar configuration option [here](https://www.packer.io/plugins/builders/vmware/iso#fusion_app_path).

In order to allow a user to properly configure Packer to work within their environment we think adding a similar option to the workstation driver would be ideal. Please let us know if you have any questions, or if you  are no longer interested in making the change.

**Comment by sparshev** [comment]
Hi @nywilken , long time no see. Not sure I understand the further steps - this particular change fixing the hardcoded paths for vmware driver. It's now actually generates from the found paths and could be modified further with additional config arg in a separated change - no issues here, but not sure we need to increase the complexity of this particular change.

I think additional option for vmware driver to configure those will be great - but the auto detection still should work in most cases, so I think those things are both needed and "auto" part is fixed by this particular change.

**Review by nywilken — COMMENTED** [review]
I know that  I'm calling out a super edge case here but it is a possible regression as we do not know user set ups. I would like to get this change in but would like to get your thoughts on the proposed change and potential PR.

**Comment by tenthirtyam** [comment]
Suggest we mark as draft until open concerns are resolved and ready for review.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Pull request closed without merging**
